Standardize the way to generate ActivityLogEntry or log message #6504

xpdavid opened the pull request:
Fixes #6504

1. Apply the builder pattern to `ActivityLogEntry`

1. Apply factory pattern to `ActivityLogEntry`, the factory class is `ActivityLogGenerator` (It's a utility class, which doesn't have any dependency on `api`, idea from #6449)

There are two small changes made to the behaviour of logging:

1. If the user is not logged in using google id, the google id field in log message is always `Unknown` (It is `Unregistered` before, which is really confusing)

1. If use is logged in with google, the role field will automatically change to the corresponding role of the user (Admin, Instructor, Student, Unregistered). The role for automated action is always Auto. If the user is logged in with auth regkey, the role will become Unregistered : [course_id].

Note that some of the magic strings in `ActivityLogEntry` for the UI part remain unchanged as they will be totally removed in #6524

wkurniawan07 commented:
@xpdavid, just confirming, is this change dependent on #6524?

xpdavid commented:
No. The two are independent. :)

wkurniawan07 commented:
@xpdavid well done then :+1: I presume this is ready for review?

xpdavid commented:
Yes. Ready for review after ec06454 passed.

xpdavid commented:
hmm, I just find a bug which may cause NPE. Pls delay the review. Sorry.

xpdavid commented:
Ready for review @wkurniawan07

wkurniawan07 commented:
@leeyimin are you going to be bored of reviewing another PR of this topic again? :joy:

wkurniawan07 commented:
@leeyimin I'm going to operate on "no response is a positive response" basis :smirk:

xpdavid commented:
@leeyimin I will do some changes first. You may delay the review. :)

xpdavid commented:
@leeyimin Ready for review.
Sorry to bother you again. To facilitate your reviewing, here are the major changes
- Add a builder to activityLogEntry and change the variable name in activityLogEntry to something more clear (e.g. `googleId` -> `userGoogleId`)

- Add a factory `ActivityLogGenerator` and provide five methods (all with details javadoc):
    - `generateNormalPageActionLogMessage`
    - `generateServletActionFailureLogMessage`
    - `generateSystemErrorReportLogMessage`
    - `generateBasicActivityLogMessage`
    - `generateActivityLogFromAppLogLine`

- Test cases for `ActivityLogEntry` and `ActivityLogGenerator` updated accordingly.

leeyimin commented on src/main/java/teammates/common/util/ActivityLogGenerator.java:
```diff
@@ -0,0 +1,336 @@
+     */
+    public String generateBasicActivityLogMessage(String url, Map<String, String[]> params,
+                                                  String message, UserType currUser) {
+        ActivityLogEntry.Builder builder = generateBasicActivityLogEntry(url, params, null);
```
Is the currUser field supposed to be currUser?

leeyimin commented on src/main/java/teammates/common/util/ActivityLogGenerator.java:
```diff
@@ -0,0 +1,336 @@
+     * 
+     * @param url URL of the request
+     * @param params parameterMap of the request
+     * @param req HttpServletRequest in the request
```
This line should be removed.

leeyimin commented on src/main/java/teammates/common/util/ActivityLogGenerator.java:
```diff
@@ -0,0 +1,336 @@
+        String servletName = getActionNameFromUrl(url);
+        long currTime = System.currentTimeMillis();
+        ActivityLogEntry.Builder builder = new ActivityLogEntry.Builder(servletName, url, currTime);
+        builder.withActionName(servletName); // by default
```
Hmm.. What do you mean by "by default"? Why call this method when the field is set by the constructor too?

xpdavid replied:
so basically these are two fields in the log message. There actionName and servletName.
If the action is normal (without exception or system error), there are the same.
If the action produces error or exception, the action will be change to something like 'Servlet Action Error' or 'System Error report'.

I should clarify this into Javadoc. :-P

leeyimin commented on src/main/java/teammates/common/util/ActivityLogGenerator.java:
```diff
@@ -0,0 +1,336 @@
+ * 
+ * @see {@link ActivityLogEntry}
+ */
+public class ActivityLogGenerator {
```
Methods in the class can be better organised by putting similarly named methods together. Can consider making the class static to avoid constructing a new generator object every time it is used by other classes.

xpdavid replied:
I put methods together if there are related (e.g. one method is used in another method). If this is confusing, I can change. :)

leeyimin commented on src/main/java/teammates/common/util/ActivityLogGenerator.java:
```diff
@@ -0,0 +1,336 @@
+        try {
+            long actionTimeTaken = tokens.length == ActivityLogEntry.POSITION_OF_LOG_TIMETAKEN + 1
+                                            ? Long.parseLong(tokens[ActivityLogEntry.POSITION_OF_LOG_TIMETAKEN].trim())
+                                            : 0;
```
Align '?' and ':' with '='.

xpdavid commented:
@leeyimin Ready for review. 
I have renamed `actionServletName` to `actionName`, `actionName` to `actionResponse` to avoid confusion.
Basically, by default (I have updated the Javadoc also), `actionResponse` and `actionName` will have the same value. However, if there is system error or exception, `actionResponse` will be changed to something like `Action Failure` `System Error` etc.
I think this is more clear than the `actionName` `actionServletName` things.

leeyimin commented on src/test/java/teammates/test/cases/util/ActivityLogGeneratorTest.java:
```diff
@@ -0,0 +1,275 @@
+        url = Const.ActionURIs.STUDENT_COURSE_JOIN;
+        mockParamMap = generateMockParamsWithRegKey();
+        logMessage = "TEAMMATESLOG|||studentCourseJoin|||studentCourseJoin|||true|||Unknown|||Unknown|||"
+                            + "Unknown|||Unknown|||Not authorized|||/page/studentCourseJoin";
```
Align the '+' to the start of the string.

leeyimin commented on src/test/java/teammates/test/cases/util/ActivityLogGeneratorTest.java:
```diff
@@ -0,0 +1,275 @@
+        Map<String, String[]> mockParamMap = Maps.newHashMap();
+        Exception e = new PageNotFoundException("randomPage");
+        String logMessagePrefix = "TEAMMATESLOG|||Error when getting ActionName for requestUrl : /randomPage"
+                + "|||Servlet Action Failure|||true|||Unregistered|||Unknown|||googleIdABC|||Unknown|||";
```
Maybe align this too to be consistent?

leeyimin commented on src/test/java/teammates/test/cases/util/ActivityLogGeneratorTest.java:
```diff
@@ -0,0 +1,275 @@
+        e = new UnauthorizedAccessException("Unknown Registration Key KeyABC");
+        generatedMessage = logCenter.generateServletActionFailureLogMessage(url, mockParamMap, e, null);
+        logMessagePrefix = "TEAMMATESLOG|||studentCourseJoin|||Servlet Action Failure|||true"
+                    + "|||Unknown|||Unknown|||Unknown|||Unknown|||";
```
Same here.

leeyimin commented on src/test/java/teammates/test/cases/util/ActivityLogGeneratorTest.java:
```diff
@@ -0,0 +1,275 @@
+        String url = Const.ActionURIs.AUTOMATED_FEEDBACK_CLOSED_REMINDERS;
+        Map<String, String[]> mockParamMap = Maps.newHashMap();
+        String logMessage = "TEAMMATESLOG|||feedbackSessionClosedReminders|||feedbackSessionClosedReminders|||true"
+                + "|||Auto|||Unknown|||Unknown|||Unknown|||auto task|||/auto/feedbackSessionClosedReminders";
```
And here.

leeyimin commented on src/test/java/teammates/test/cases/util/ActivityLogGeneratorTest.java:
```diff
@@ -0,0 +1,275 @@
+        
+        url = Const.ActionURIs.STUDENT_COURSE_JOIN + "?user=test@email.com&course=1";
+        logMessage = "TEAMMATESLOG|||studentCourseJoin|||studentCourseJoin|||true|||Unregistered:CS2103|||Joe"
+                            + "|||Unknown|||student@email|||Join Course|||" + url;
```
Same here.

leeyimin commented on src/test/java/teammates/test/cases/util/ActivityLogGeneratorTest.java:
```diff
@@ -0,0 +1,275 @@
+        url = Const.ActionURIs.STUDENT_HOME_PAGE + "?course=A&user=test";
+        mockParamMap = Maps.newHashMap();
+        logMessage = "TEAMMATESLOG|||studentHomePage|||studentHomePage|||true|||Unregistered|||Unknown"
+                + "|||googleId|||Unknown|||Try student home|||" + url;
```
Align the '+' to the start of the string.

leeyimin commented on src/test/java/teammates/test/cases/util/ActivityLogGeneratorTest.java:
```diff
@@ -0,0 +1,275 @@
+        ______TS("Google login (Insturctor)");
+        
+        String logTemplate = "TEAMMATESLOG|||%1$s|||%1$s|||true|||%2$s|||david"
+                + "|||googleId|||david@email.com|||View Result|||/page/%1$s";
```
Align the '+' to the start of the string.

leeyimin commented on src/test/java/teammates/test/cases/util/ActivityLogGeneratorTest.java:
```diff
@@ -0,0 +1,275 @@
+        
+        url = Const.ActionURIs.ADMIN_ACTIVITY_LOG_PAGE;
+        logMessage = "TEAMMATESLOG|||adminActivityLogPage|||adminActivityLogPage|||true|||Admin|||david"
+                + "|||googleId|||david@email.com|||View Result|||/admin/adminActivityLogPage";
```
Same here.

leeyimin commented on src/test/java/teammates/test/cases/util/ActivityLogGeneratorTest.java:
```diff
@@ -0,0 +1,275 @@
+        userType.isAdmin = true;
+        acc = new AccountAttributes("anotherGoogleId", "david", false, "david@email.com", "NUS");
+        logMessage = "TEAMMATESLOG|||instructorCommentsPage|||instructorCommentsPage|||true|||Instructor(M)|||david"
+                + "|||anotherGoogleId|||david@email.com|||View comments|||/page/instructorCommentsPage";
```
Same here.

leeyimin commented on src/test/java/teammates/test/cases/util/ActivityLogGeneratorTest.java:
```diff
@@ -0,0 +1,275 @@
+    public void testGenerateActivityLogFromAppLogLine() {
+        ______TS("Success: Generate activityLog from appLogLine (with TimeTaken)");
+        String logMessageWithoutTimeTaken = "TEAMMATESLOG|||instructorHome|||Pageload|||true|||Instructor"
+                    + "|||UserName|||UserId|||UserEmail|||Message|||URL|||UserId20151019143729608";
```
Align the '+' to the start of the string.

leeyimin commented on src/test/java/teammates/test/cases/util/ActivityLogGeneratorTest.java:
```diff
@@ -0,0 +1,275 @@
+                                  + "|||Unknown|||googleIdABC|||Unknown|||";
+        
+        String generatedMessage = logCenter.generateSystemErrorReportLogMessage(url, mockParamMap,
+                                                                errorEmail, loginUser);
```
Inconsistent indentation.

leeyimin commented on src/test/java/teammates/test/cases/util/ActivityLogGeneratorTest.java:
```diff
@@ -0,0 +1,275 @@
+        errorEmail = generateMockEmailWrapperFromRequest(e);
+        
+        generatedMessage = logCenter.generateSystemErrorReportLogMessage(url, mockParamMap,
+                                                                    errorEmail, null);
```
Inconsistent indentation.

wkurniawan07 commented on src/main/java/teammates/common/util/ActivityLogEntry.java:
```diff
@@ -1,359 +1,243 @@
-    private String logInfoAsHtml;
+    private long actionTimeTaken;
+    
+    // legacy of messing up UI and logic
```
What does this mean? 🤔

xpdavid replied:
You can see there is a magic field in the log message.
`TEAMMATESLOG|||...|||true|||...`. This field is actually true for all logs.
Previously they use this field to control the displaying of activitlyLog in `AdminActivitLogPageAction`. It is just UI thing, not relevant to the log itself.

The thing is that we cannot remove it for compatibility reason.

wkurniawan07 commented on src/main/java/teammates/common/util/ActivityLogEntry.java:
```diff
@@ -607,25 +372,126 @@ public void highlightKeyStringInMessageInfoHtml() {
+     * Field actionResponse will be initialized to have the same value as actionName.
+     * All null values (if possible) that are passed into the builder will be ignored.
+     * 
+     * @see {@link ActivityLogEntry}
```
As mentioned in #6583, `@see` attaches link to the entry automatically so there is no need to `@link` it.

wkurniawan07 commented on src/main/java/teammates/common/util/ActivityLogEntry.java:
```diff
@@ -607,25 +372,126 @@ public void highlightKeyStringInMessageInfoHtml() {
+            return logTime;
+        }
+        
+        public String getActionName() {
```
Why would we need getters for a builder class?

xpdavid replied:
As I don't want to pass the `actionName` everywhere in methods. `actionName` and `logTime` are required fields and cannot be changed once builder is created.

wkurniawan07 commented on src/main/java/teammates/common/util/ActivityLogGenerator.java:
```diff
@@ -0,0 +1,329 @@
+        }
+    }
+
+    private void degradeRoleToStudentIfNecessary(Builder builder) {
```
`degrade` -> `downgrade`

wkurniawan07 commented on src/main/java/teammates/common/util/ActivityLogGenerator.java:
```diff
@@ -0,0 +1,329 @@
+        }
+    }
+
+    private void degradeRoleToInstructorIfNecessary(Builder builder) {
```
`degrade` -> `downgrade`

wkurniawan07 commented on src/main/java/teammates/common/util/ActivityLogGenerator.java:
```diff
@@ -0,0 +1,329 @@
+    private boolean isInstructorPage(String actionName) {
+        return actionName.toLowerCase().startsWith(Const.ActivityLog.PREFIX_INSTRUCTOR_PAGE)
+                || Const.ActionURIs.INSTRUCTOR_FEEDBACK_STATS_PAGE.contains(actionName);
+        // TODO rename the special INSTRUCTOR_FEEDBACK_STATS_PAGE to start with instructor
```
This does not directly concern logs. If you want to add this TODO, don't add it here.

wkurniawan07 commented on src/main/java/teammates/common/util/ActivityLogGenerator.java:
```diff
@@ -0,0 +1,329 @@
+    private ActivityLogEntry initActivityLogUsingAppLogMessage(AppLogLine appLog, String[] tokens) {
+        // TEAMMATESLOG|||ACTION_NAME|||ACTION_RESPONSE|||TO_SHOW|||ROLE|||NAME|||GOOGLE_ID|||EMAIL
+        // |||MESSAGE(IN HTML)|||URL|||TIME_TAKEN
+        String actionName = tokens[ActivityLogEntry.POSITION_OF_ACTION_NAME];
```
I'm quite wary of the need to expose `POSITION_OF_*` variable. Can't we just construct an `ActivityLogEntry` object with `AppLogLine` as constructor parameter, similar to how it was done before?

xpdavid replied:
I thought we could put everything to one place and make it consistent.

I think we should not expose `POSITION_OF_*` variable, I will change later.

wkurniawan07 commented on src/test/java/teammates/test/cases/util/ActivityLogGeneratorTest.java:
```diff
@@ -0,0 +1,275 @@
+    
+    private ActivityLogGenerator logCenter = new ActivityLogGenerator();
+
+    // TODO: can move log verification in action test cases to here or remove them totally
```
Try not to sprinkle too much TODOs everywhere. They're easily forgotten and you've seen how much we have in the code base right now.

wkurniawan07 commented on src/test/java/teammates/test/cases/util/ActivityLogGeneratorTest.java:
```diff
@@ -0,0 +1,275 @@
+        UserType loginUser = new UserType("googleIdABC");
+        Map<String, String[]> mockParamMap = Maps.newHashMap();
+        String url = Const.ActionURIs.INSTRUCTOR_HOME_PAGE;
+        @SuppressWarnings("PMD.AvoidThrowingNullPointerException") Exception e = new NullPointerException();
```
Do we really need `e` to be NPE here?

wkurniawan07 commented on src/test/java/teammates/test/cases/util/ActivityLogGeneratorTest.java:
```diff
@@ -0,0 +1,275 @@
+        EmailWrapper errorReport =
+                new EmailGenerator().generateSystemErrorEmail("GET", "MAC OS", "/page/somePage",
+                                                              "http://example/", "{}", null, t);
+        return errorReport;
```
Two things here:
- This is not really a mock, isn't it? It's calling the actual email-generating function.
- Simply `return new EmailGenerator()....` will make it shorter.

xpdavid commented:
Delay the review of this until #6524 is merged.

xpdavid commented:
Finally, I find that I cannot manage this PR anymore as there are too many conflict after #6524. Going to close this and create a new one. :P
